lib/cargo.js

```diff
@@ -82,65 +82,131 @@ export function provideBuilder() {
+            severity: level2severity(json.level),
+            trace: trace
+          });
+        });
+      }
+
+      function matchFunction(output) {
+        const useJson = atom.config.get('build-cargo.jsonErrors');
+        const messages = [];  // resulting collection of high-level messages
+        let msg = null;       // current high-level message (error, warning or panic)
+        let sub = null;       // current submessage (note or help)
+        let match = null;
+
+        output.split(/\n/).forEach(line => {
+          // Cargo final error messages start with 'error:', skip them
+          if (line === null || line === '' || line.substring(0, 6) === 'error:') {
+            msg = null;
+            sub = null;
+          } else if (useJson && line[0] === '{') {
+            // Parse a JSON block
+            parseJsonOutput(line, messages);
+          } else if (line.substring(0, 16) === 'stack backtrace:') {
+            // Start a stacktrace submessage (it's always under a panic message)
+            if (msg !== null) {
+              sub = {
+                message: 'Stack backtrace',
+                type: 'Stack'
+              };
+              msg.trace.push(sub);
+            }
+          } else {
+            // Check for compilation messages
+            match = /^(.+.rs):(\d+):(\d+):(?: (\d+):(\d+))? (error|warning|help|note): (.*)/g.exec(line);
+            if (match) {
+              const type = match[6];
+              const message = match[7];
+              if (type === 'error' || type === 'warning' || msg === null) {
+                msg = {
+                  message: message,
+                  file: match[1],
+                  line: match[2],
+                  line_end: match[4],
+                  col: match[3],
+                  col_end: match[5],
+                  type: type,
+                  trace: []
+                };
+                messages.push(msg);
+                sub = null;
+              } else {
+                sub = {
+                  message: message,
+                  type: 'info'
+                };
+                msg.trace.push(sub);
+              }
+            } else {
+              // Check for panic
+              match = /^(thread '[^']+' panicked at '[^']+'), ([^\/][^\:]+):(\d+)/g.exec(line);
+              if (match) {
+                msg = {
+                  message: match[1],
+                  file: match[2],
+                  line: match[3],
+                  type: 'error',
+                  trace: []
+                };
+                messages.push(msg);
+                sub = null;
+              } else {
+                // Just a description in the current block. Only add it when in submessage
+                // because Linter does the job for high-level messages.
+                if (sub !== null) {
```

Comment:
The reason I have this additional nesting level is that otherwise I would have to either make two matches on the previous level (one for error/warning/note/help, another for panic) which might slightly affect performance or use `else if ((match = regex.match(...))` that would work, but is quite error-prone for future maintenance (and probably for static analysis tools), so I decided to live it the way I did.
But I can make it more flat, that's not a problem at all.